framework/include/base/SubProblem.h

```diff
@@ -368,6 +400,12 @@ class SubProblem : public Problem
   /* This needs to remain <unsigned int> for threading purposes */
   std::vector<unsigned int> _has_active_elemental_moose_variables;
 
+  /// Set of material property ids that determine whether materials get reinited
+  std::vector<std::set<unsigned int>> _active_material_property_ids;
+
+  /// True if _active_material_property_ids is not empty
+  std::vector<unsigned int> _has_active_material_property_ids;
```

Comment:
> There are certain members of the team that are terrified and completely against vector because it has bitten them (us) before in threaded cases. 

You say this like my dislike of `std::vector<bool>` is unjustified... but it's not.  Indeed, in this very case: if you were to use a `std::vector<bool>` you would have a thread race condition that would lead to incorrect execution!

There are very real and VERY good reasons for avoiding `std::vector<bool>` in a code that relies on threading.  There is absolutely no reason to use it until you can justify that the memory savings is worth it... and then that variable should have a ton of comments around it telling people what's up.